internal/goldendataset/span_generator.go

```diff
@@ -0,0 +1,559 @@
+	statusMap[SpanStatusCancelled] = "Cancellation received"
+	statusMap[SpanStatusUnknownError] = ""
+	statusMap[SpanStatusInvalidArgument] = "parameter is required"
+	statusMap[SpanStatusDeadlineExceeded] = "timed out after 30002 ms"
+	statusMap[SpanStatusNotFound] = "/dragons/RomanianLonghorn not found"
+	statusMap[SpanStatusAlreadyExists] = "/dragons/Drogon already exists"
+	statusMap[SpanStatusPermissionDenied] = "tlannister does not have write permission"
+	statusMap[SpanStatusResourceExhausted] = "ResourceExhausted"
+	statusMap[SpanStatusFailedPrecondition] = "33a64df551425fcc55e4d42a148795d9f25f89d4 has been edited"
+	statusMap[SpanStatusAborted] = ""
+	statusMap[SpanStatusOutOfRange] = "Range Not Satisfiable"
+	statusMap[SpanStatusUnimplemented] = "Unimplemented"
+	statusMap[SpanStatusInternalError] = "java.lang.NullPointerException"
+	statusMap[SpanStatusUnavailable] = "RecommendationService is currently unavailable"
+	statusMap[SpanStatusDataLoss] = ""
+	statusMap[SpanStatusUnauthenticated] = "nstark is unknown user"
+	return statusMap
+}
+
+func GenerateSpans(count int, startPos int, pictFile string) ([]*otlptrace.Span, int, error) {
+	pairsData, err := loadPictOutputFile(pictFile)
+	if err != nil {
+		return nil, 0, err
+	}
+	pairsTotal := len(pairsData)
+	spanList := make([]*otlptrace.Span, count)
+	index := startPos + 1
+	var inputs []string
+	traceID := generateTraceID()
+	parentID := generateSpanID()
+	for i := 0; i < count; i++ {
+		inputs = pairsData[index]
+		if SpanParentRoot == inputs[ColumnParent] {
+			traceID = generateTraceID()
+			parentID = nil
+		}
+		spanName := generateSpanName(inputs[ColumnAttributes], i)
+		spanList[i] = GenerateSpan(traceID, inputs[ColumnTracestate], parentID, spanName, inputs[ColumnKind],
+			inputs[ColumnAttributes], inputs[ColumnEvents], inputs[ColumnLinks], inputs[ColumnStatus])
+		parentID = spanList[i].SpanId
+		index++
+		if index >= pairsTotal {
+			index = 1
+		}
+	}
+	return spanList, index, nil
+}
+
+func loadPictOutputFile(fileName string) ([][]string, error) {
+	file, err := os.Open(filepath.Clean(fileName))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	reader := csv.NewReader(file)
+	reader.Comma = '\t'
+
+	return reader.ReadAll()
+}
+
+func generateSpanName(spanTypeID string, index int) string {
+	if SpanAttrHTTPClient == spanTypeID {
+		return fmt.Sprintf("/dragons/%d", index)
+	} else if SpanAttrHTTPServer == spanTypeID {
+		return "/dragons/{dragonId}"
+	} else if SpanAttrGRPCClient == spanTypeID || SpanAttrGRPCServer == spanTypeID {
+		return "com.example.PetFoodService/DispenseFeed"
+	} else {
+		return fmt.Sprintf("gotest%d", index)
+	}
+}
+
+func GenerateSpan(traceID []byte, tracestate string, parentID []byte, spanName string, kind string, spanTypeID string,
+	eventCnt string, linkCnt string, statusStr string) *otlptrace.Span {
+	endTime := time.Now().Add(-50 * time.Microsecond)
```

Comment:
Nit: `time.Sub` may be slightly easier to understand than `time.Add` of negative duration.